src/encoding.rs

```diff
@@ -716,6 +716,44 @@ macro_rules! length_delimited {
 pub mod string {
     use super::*;
 
+    // String::as_mut_vec is unsafe because it doesn't check that the bytes
+    // inserted into the resulting vec are valid UTF-8. We check
+    // after fully extending the buffer in order to ensure this is safe.
+    // If an unwind happens before the check is performed and the result
+    // is committed, the string is truncated to the previously valid length.
+    struct StringGuard<'a> {
+        string: &'a mut String,
+        committed_len: usize,
+    }
+
+    impl<'a> Drop for StringGuard<'a> {
+        fn drop(&mut self) {
+            self.string.truncate(self.committed_len);
+        }
+    }
+
+    impl<'a> StringGuard<'a> {
+        fn new(string: &'a mut String) -> Self {
+            let committed_len = string.len();
+            StringGuard {
+                string,
+                committed_len,
+            }
+        }
+
+        unsafe fn as_mut_vec(&mut self) -> &mut Vec<u8> {
+            self.string.as_mut_vec()
+        }
+
+        fn commit(mut self) -> Result<(), DecodeError> {
+            let new_bytes = &self.string.as_bytes()[self.committed_len..];
+            str::from_utf8(new_bytes)
```

Comment:
In a general purpose `StringGuard` I would do, but here we can use the knowledge that `bytes::merge` would not change the already filled part of the vector, and nothing else can modify it while it is borrowed.

But again, clearing the string value before the new wire data are merged, and also on `StringGuard` drop without commit, would be simpler and not rely on the other function's behavior to be safe. I just don't understand why this implementation concatenates non-repeating fields of types `string` and `bytes`.

Reply:
> clearing the string value before the new wire data are merged... would be simpler

Also in the `BytesString` (or `StrChunk`) rework in #190, I would just suck the data into `Bytes` and convert that into the new string value with a safe UTF-8 content check rather than fiddle with unsafe mutable representations.

Reply:
>  I just don't understand why this implementation concatenates non-repeating fields of types string and bytes.

It's entirely possible that this is done by accident, and should be fixed.  Do you have a reference for where the spec / PB docs mention that merging will overwrite strings?

Reply:
In the [developer guide on encoding](https://developers.google.com/protocol-buffers/docs/encoding#optional):

> Normally, an encoded message would never have more than one instance of a non-repeated field. However, parsers are expected to handle the case in which they do. For numeric types and strings, if the same field appears multiple times, the parser accepts the _last_ value it sees.